### PR TITLE
Filter fixes

### PR DESCRIPTION
**Pull request by etpinard**
After first being released in `v1.18.0`, we come to the conclusion that filter transforms are incompatible with plot.ly's webapp data references.

So, we have no choice but to :hocho: the `filtersrc` attribute and rename it `'target'`. A backward-compatible block has been to `cleanData`.

Moreover, to allow filtering about arbitrary columns of data, we now allow `'target'` to be an arbitrary array of data - which required a few patches to the `Lib.findArrayAttributes`.

cc @rreusser @bpostlethwaite @chriddyp

### CODE REVIEW

File: src/lib/coerce.js [hunk 1]
```diff
@@ -458,6 +458,17 @@ exports.findArrayAttributes = function(trace) {
...
     exports.crawl(trace._module.attributes, callback);
 
+    if(trace.transforms) {
+        var transforms = trace.transforms;
+
+        for(var i = 0; i < transforms.length; i++) {
+            var transform = transforms[i];
+
+            stack = ['transforms[' + i + ']'];
+            exports.crawl(transform._module.attributes, callback, 1);
```
Comment (etpinard) [inline src/lib/coerce.js]: Loop through transform container found in the trace, use the transform module attributes declaration to determine which attributes are `data_array` or `arrayOk`.

File: src/transforms/filter.js [hunk 2]
```diff
@@ -157,18 +161,45 @@ exports.calcTransform = function(gd, trace, opts) {
...
+    // In the case of an array target, make a mock data array
+    // and call supplyDefaults to the data type and
+    // setup the data-to-calc method.
+    if(Array.isArray(target)) {
+        var mockGd = {
+            data: [{ x: target }],
+            layout: {}
+        };
+
+        Plots.supplyDefaults(mockGd);
```
Comment (etpinard) [inline src/transforms/filter.js]: Calling the full `supplyDefaults` here is an overkill.

@monfera factored the autoType routine inhttps://github.com/plotly/plotly.js/pull/1021/commits/5b71b47d0ed1d437708da2a293481e911828865a, I might just use that instead once #1021 is merged.

**Comment by etpinard** [comment]
Oh and by the way @bpostlethwaite transform attribute are already part of the plot schema:

image

**Comment by bpostlethwaite** [comment]
:+1:! 
`targetsrc` doesn't need to be part of Plotly.js. I think we can safely look at removing all the `*src` keys from Plotlyjs attributes!

:dancer: from me!

**Pull request merged**